### ISSUE TITLE
Introduce a replacement for `gradle.allprojects {...}` suitable for incremental adoption

### ISSUE THREAD

**Issue by bamboo** [gradle/gradle#29187]
The new API should preserve the order of configuration for each project (but not necessarily across projects).

For example, in the following scenario, the order in which Gradle runs the configuration logic associated with subproject `:a` would be preserved by the new API when compared to the order in which `gradle.allprojects` and `project.subprojects` run them.
 
```kotlin 
// settings.gradle.kts
rootProject.name = "root"
include("a")
gradle.lifecycle.allprojectsReplacement {
   println("allprojects:$name!")
}

// build.gradle.kts
println("root!")
subprojects {
  println("sub:$name!")
}

// a/build.gradle.kts
println("a!") 
```

The expected output would be:

```
allprojects:root!
root!
allprojects:a!
sub:a!
a!
```

Which preserves the order in which `a` is configured compared to the `gradle.allprojects` output:
```
allprojects:root!
allprojects:a!
root!
sub:a!
a!
```

**Comment by bamboo** [comment]
Related to #29013 (Develocity lifecycle service should be migrated to the new API once ready).

**Comment by bamboo** [comment]
Implemented by [#29915](https://github.com/gradle/gradle/pull/29915)
